### ISSUE TITLE
Incorrect marking of code as 'not executed' following merging of cov files

### ISSUE THREAD

**Issue by iainfogg** [sebastianbergmann/php-code-coverage#409]
Hi, I'm running using Robo (http://robo.li/) to run a series of Codeception unit tests which are then having the .cov files merged (using phpcov) and an HTML file produced to show code coverage - I mention all the steps in case that's relevant.

However, I'm finding that quite a few lines are being marked as 'not executed' when, depending on your perspective, they either were executed or are dead code (unless I'm misunderstanding how this should work). For example, the first brace here - also, I don't know if this is the correct marking for the closing brace.

![screen shot 2015-12-02 at 13 42 37](https://cloud.githubusercontent.com/assets/7030201/11532323/160e8f3a-98fb-11e5-8d98-29948edbc020.png)

Also, in this one, how can line 52 be not executed?

![screen shot 2015-12-02 at 13 49 38](https://cloud.githubusercontent.com/assets/7030201/11532422/9c3a2718-98fb-11e5-85f4-0623d51eb500.png)

Or in this one, how can line 54 be not executed code?

![screen shot 2015-12-02 at 13 50 22](https://cloud.githubusercontent.com/assets/7030201/11532443/b94361bc-98fb-11e5-97d3-94d79988abce.png)

My composer.lock file shows php-code-coverage at version 2.2.4, and phpcov at 2.0.2. phpUnit is at 4.8.19 and codeception is at 2.1.4.

Any ideas on what the issue is, or anything I can do to help get to the bottom of it?

Thanks, Iain

**Comment by iainfogg** [comment]
PS I just noticed that line 52 appears to be the same issue as here https://github.com/sebastianbergmann/php-code-coverage/issues/406

**Comment by iainfogg** [comment]
After a bit more tinkering, it appears that this may be related to the merging. Here's a result from running the test directly from codeception:

![screen shot 2015-12-02 at 14 49 13](https://cloud.githubusercontent.com/assets/7030201/11533895/ec5c4110-9903-11e5-9c37-76793e644808.png)

Here's the same code from a report which was produced by using robo to run two separate sets unit tests then merge the resulting cov files to produce an html file:

![screen shot 2015-12-02 at 14 50 53](https://cloud.githubusercontent.com/assets/7030201/11533942/1ed6bfb2-9904-11e5-8f6f-47585d3aaa1c.png)

The first shows what I would expect; the second seems wrong.

Hope the extra info helps.

**Comment by iainfogg** [comment]
OK, here's a progress update on this. It is related to the merging that's happening. The issue is around how the list of lines in files that have not been called, but are in the whitelisted folders, are added in.

getLinesToBeIgnored is called, and it trawls through the code identifying tokens, and looking for those lines that should be ignored. This processing is producing a different list of lines to the code that works out which lines are executable for code which has actually run.

When the two cov files are merged together, the representation of the file which has not run contains extra lines, and these then all appear as unexecuted lines in the merged file. That's why line 65 in the earlier example is marked red - line 65 didn't exist in the cov file from when that file was run, but in another test run which never called it, it was added as a non-run whitelisted file and line 65 was added to that cov.

One issue that I have identified is below - it was checking for PHP_Token_OPEN_TAG and CLOSE_TAG, but when I ran through with a debugger, it identified the opening and closing braces of a method as PHP_Token_OPEN_CURLY and CLOSE_CURLY - by amending the case statement as below (I wasn't sure if TAG can ever occur, so left it in just in case) the opening and closing brackets in classes that were run in one test run, but not in another test run, now appear correctly in the merged result.

![screen shot 2015-12-03 at 13 59 22](https://cloud.githubusercontent.com/assets/7030201/11562476/1618795c-99c6-11e5-80b1-8c065cc17a43.png)

However, that has also introduced an issue where closing brackets in the middle of a function e.g. on the end of an if () {} will now also appear with no colouring in classes, regardless of whether they were run in any test or not. Ideally, if the class had been called, the merge process keep the marking indicating whether they were run, or were dead code, rather than defaulting to not having an entry at all for them.

I'll keep digging, but hopefully someone who understands this rather better than me can pitch in!

Cheers, Iain

**Comment by iainfogg** [comment]
OK, I have some insight into what's gone on here. The issue is related to the merging of cov files prior to producing the report.

For files which were not executed, but were required to appear in the HTML report, the list of lines is generated, but by a relatively inaccurate assessment of which lines are executable and should therefore be monitored. Essentially, this includes all the lines from the opening bracket in a method up to and including the closing bracket of each method.

For files which are actually executed, as I understand it they are analysed by Xdebug to get a definitive list of executable lines in that file. This does not include the opening bracket of each method, and doesn't include certain other lines.

The issue is coming because the two approaches are generating a different list of lines - the non-executed code has more lines in it than the executed code.

When a single run is being done, this is never an issue; however, when two cov files are being merged, and the two runs include a class which was executed in one run / cov file, but not in the other run / cov file, a bunch of lines (the ones which only appear in the non-executed code) get added into the merged file but can never be marked as having been executed, thereby throwing the stats for that particular class.

I have a solution which seems to be working for me, although there may be a better approach that someone can suggest? However, the idea is that if the number of lines differ between two cov files for a particular file, always accept by default the one with the fewest lines (as that must be the executed one, which will have the most accurate list of lines that can be run). Doing so avoids tagging lines incorrectly as not executed in the final report.

Does this seem a sensible approach? If so, I can do a PR for it.

I've added this code in the merge method, just before going through and doing the line by line merging of results, to avoid that if there are lines and they don't match in number:

```
            // Check if there are some lines already present in the two
            // versions being merged, and also that the number of lines
            // in each is different. This indicates that one is based on
            // pulling lines in from glancing through the code in unexecuted
            // code (the one with more lines in), and one is based on using Xdebug
            // to properly analyse the code (the one with less lines in).
            // In this scenario, ditch the lines from the unexecuted analysis,
            // as it will contain lines which are not classed as executable by
            // Xdebug, and those lines will throw the code analysis by appearing
            // as if they should be run but weren't.
            if ((count($lines) > 0)
                && (count($this->data[$file]) > 0)
                && (count($lines) != count($this->data[$file]))
            ) {
                if (count($lines) > count($this->data[$file])) {
                    // More lines in the one being added in, so just ignore them
                    // and move on to the next file.
                    continue;
                } else {
                    // More lines in the one we currently have, so ditch what we
                    // currently have and take what's being merged in exactly as
                    // it is.
                    $this->data[$file] = $lines;

                    // Move on to the next file.
                    continue;
                }
            }
```

Thanks, Iain

**Comment by dunglas** [comment]
Hello,

I'm having the exact same issue. If you agree to include @iainfogg's fix, I can open the PR.

Thanks.

**Comment by KintradimCrux** [comment]
Hello,

I'm having the exact same issue - running only one test class shows 100% of executable lines are covered, while when running all tests the same class shows random "not executed" lines that either contain comments or definitely were executed. I just ran the tests with @iainfogg's fix and it now shows the correct coverage.

**Comment by sebastianbergmann** [comment]
@iainfogg Please send a pull request, thanks.

**Comment by iainfogg** [comment]
Apologies to @dunglas - I never saw his comment from August 2016.

I have added the PR as requested https://github.com/sebastianbergmann/php-code-coverage/pull/542

However, that pull request is based on the work I did 18 months ago which was in branch 2.2. Obviously that's pretty old now; maybe someone can either advise how to get this into later branches too if needed, or consider whether that fix is likely to be compatible with later branches, as I have not looked at anything later.

It's my first PR, so be gentle with me :)

Thanks @sebastianbergmann

**Comment by KintradimCrux** [comment]
@iainfogg I inserted your code sample inside the outer foreach loop of the method merge() just before the inner foreach loop is run, it seems to work with both phpunit 6.2.3 and 5.7.21.

**Comment by iainfogg** [comment]
So should I be looking at adding this to every branch (2.2, 3.0, 3.1, 3.2, 3.3, 4.0, 5.0 and master)? And is it a separate PR for each one? Or should I just add it to master? @KintradimCrux

**Comment by sebastianbergmann** [comment]
As you can see on at https://phpunit.de/build-status.html, php-code-coverage 4.0 is the oldest, still supported branch. So please send a pull request to the 4.0 branch and I'll take it from there. Thanks!

**Comment by iainfogg** [comment]
@sebastianbergmann I will be doing this and coming up with a new pull request on 4.0. The code's easy to add, but the data in the tests looks like it's going to take a reasonable amount of going through so it'll pass tests and properly test the new code. I'll let you know once it's done.

**Comment by KintradimCrux** [comment]
@sebastianbergmann I only tested this with version 4.0.8 and 5.1.2 of the php-code-coverage package, i cannot tell if this works with older versions

**Comment by pixelcat** [comment]
Gentle bump on this one. Looks like the build on this failed for an unrelated reason.

**Comment by iainfogg** [comment]
Sorry @pixelcat I am struggling to get the time to fix the failed build, as I don't understand enough of what's making that other test fail. If anyone else can pick this up so this can get implemented, that would be great.

**Comment by KintradimCrux** [comment]
@iainfogg I just found that this patch solves the problem for many files in my bundle project, by sorting out the not executable lines. But for one class it lowers code coverage to 0 since this class is used in many of the classes to test and the test cases covering it not merged. I'm currently testing how to work around this issue.

**Comment by KintradimCrux** [comment]
@iainfogg This replacement code does the trick:

```php
            // Check if there are some lines already present in the two
            // versions being merged, and also that the number of lines
            // in each is different. This indicates that one is based on
            // pulling lines in from glancing through the code in unexecuted
            // code (the one with more lines in), and one is based on using Xdebug
            // to properly analyse the code (the one with less lines in).
            // In this scenario, ditch the additional lines in the larger array,
            // then the results are merged.
            if ((count($lines) > 0)
            && (count($this->data[$file]) > 0)
            && (count($lines) != count($this->data[$file]))
            ) {
            	if (count($lines) > count($ours = $this->data[$file])) {
            		// More lines in the one being added in
            		$lines = array_filter(
            			$lines,
            			function($value, $key) use ($ours) {
	            			return array_key_exists($key, $ours);
	            		},
	            		ARRAY_FILTER_USE_BOTH
            		);
            	} else {
            		// More lines in the one we currently have
            		$this->data[$file] = array_filter(
            			$this->data[$file],
            			function($value, $key) use ($lines) {
	            			return array_key_exists($key, $lines);
	            		},
	            		ARRAY_FILTER_USE_BOTH
            		);
            	}
            }
```

**Comment by KintradimCrux** [comment]
@iainfogg, i meanwhile checked the patch from my last post against the php-code-coverage test suite and CodeCoverageTest::testMerge() fails, but seems to test unrealistic behaviour.

@sebastianbergmann, can you please check if CodeCoverageTest::testMerge() should test merging coverage results of parts of a file? For all tests of my bundle test suite i checked, the coverage result arrays to be merged contained the almost exact same number of lines, except for the deviations @iainfogg intended to remove with his patch.

**Comment by sebastianbergmann** [comment]
Waiting for an updated pull request.

**Comment by dekker-m** [comment]
The issue still isn't completely resolved. The pull request by @celowsky fixed a lot of the errors mentioned in this issue, but it does not handle all (or at least some) cases of dead code correctly.

The PHP coverage format seems to assign `NULL` as a value if dead code is detected. If another coverage file did not execute the file, this line is still marked as "not executed", after the results have been merged. The `array_key_exists` function in the anonymous function in `merge` returns true for lines which have been marked as dead code. If XDebug detected a line as being dead code, it should not be able to execute it during other runs. In other words: if one of the code coverage reports contains a `null` value for one of the lines, the merged results should be `null` as well. Perhaps someone can think of a where this is false, but from what I can see, this is the case.

To mark those lines as "dead code" in the merged result, there should be a check if lines are `null` in either of the source coverage files. This can easily be implemented in the last foreach loop in the `merge` method. For example:

```php
foreach ($lines as $line => $data) {
    if ($data === null || $this->data[$file][$line] === null) {
        // if the line is marked as "dead code" in either, mark it as dead code in the merged result
        $this->data[$file][$line] = null;
    } else if (!isset($this->data[$file][$line])) {
        // if no data has been set in the current data, overwrite all
        $this->data[$file][$line] = $data;
    } else {
        // otherwise merge data from both coverage files
        $this->data[$file][$line] = \array_unique(
            \array_merge($this->data[$file][$line], $data)
        );
    }
}
```

With the above code the test cases for our project yield the same result when run in a single run and when run in parallel over 4 processes. There are still some minor differences, but this does not look related to merging the coverage files. In fact, the merged .cov files are more accurate then the original coverage in some edge cases with anonymous functions.

**Comment by sebastianbergmann** [comment]
@dekker-m Please open a new ticket for this. Thanks.

**Comment by dekker-m** [comment]
Done, see #614
